rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -161,7 +161,34 @@ impl RpcState {
         &self,
         params: &serde_json::Value,
     ) -> Result<T, RpcError> {
-        Self::deserialize_call(self.rpc_call_no_deserialize(params)?.into_json().unwrap())
+        #[cfg(test)]
+        {
+            use std::collections::hash_map::DefaultHasher;
+            use std::hash::{Hash, Hasher};
+            use std::path::Path;
+
+            let dir = Path::new("test-responses");
+            let mut hasher = DefaultHasher::new();
+            params.to_string().hash(&mut hasher);
+            let hash = hasher.finish();
+            let filename = format!("{}-{}.json", self.chain, hash);
+            let file = dir.join(filename);
+
+            if file.exists() {
+                let res = std::fs::read_to_string(file).unwrap();
+                serde_json::from_str(&res).map_err(|err| RpcError::RpcCall(err.to_string()))
+            } else {
+                let res: serde_json::Value = Self::deserialize_call(
+                    self.rpc_call_no_deserialize(params)?.into_json().unwrap(),
+                )?;
+                std::fs::write(file, res.to_string()).unwrap();
+                serde_json::from_value(res).map_err(|err| RpcError::RpcCall(err.to_string()))
+            }
+        }
+        #[cfg(not(test))]
+        {
+            Self::deserialize_call(self.rpc_call_no_deserialize(params)?.into_json().unwrap())
+        }
```

Comment:
These could diverge and the "real" path would not be tested. Could you find a way to use a single build or handle the test setup outside? I think an option is to receive a reader rather than a parsed JSON, so you can choose between sending a file or the real arguments.

Reply:
a file is never sent, here a hash is calculated from the params, which are always given,
and based on it we check the existance  of a local file which if it exists contains the response to the api with said arguments.

The api call thus needs to only be made once and in CI it can be cached.